modules/apigee_edge_apidocs/src/ApiDocSpecFetcher.php

```diff
@@ -0,0 +1,264 @@
+      $request = new Request('GET', $file_uri);
+      $options = [
+        'exceptions' => TRUE,
+        'allow_redirects' => [
+          'strict' => TRUE,
+        ],
+      ];
+
+      // Generate conditional GET header.
+      if (!$apidoc->get('fetched_timestamp')->isEmpty()) {
+        $request = $request->withAddedHeader('If-Modified-Since', gmdate(DateTimePlus::RFC7231, $apidoc->get('fetched_timestamp')->value));
+      }
+
+      try {
+        $response = $this->httpClient->send($request, $options);
+
+        // In case of a 304 Not Modified there are no changes, but update
+        // last fetched timestamp.
+        if ($response->getStatusCode() == 304) {
+          $apidoc->set('fetched_timestamp', time());
+          return TRUE;
+        }
+      }
+      catch (RequestException $e) {
+        $this->log($apidoc, static::TYPE_ERROR, 'API Doc %label: Could not retrieve OpenAPI specification file located at %url.', [
+          '%url' => $file_uri,
+          '%label' => $apidoc->label(),
+        ], $show_messages);
+        return FALSE;
+      }
+
+      $data = (string) $response->getBody();
+      if (empty($data)) {
+        $this->log($apidoc, static::TYPE_ERROR, 'API Doc %label: OpenAPI specification file located at %url is empty.', [
+          '%url' => $file_uri,
+          '%label' => $apidoc->label(),
+        ], $show_messages);
+        return FALSE;
+      }
+
+      // Only save file if it hasn't been fetched previously.
+      $data_md5 = md5($data);
+      $prev_md5 = $apidoc->get('spec_md5')->isEmpty() ? NULL : $apidoc->get('spec_md5')->value;
+      if ($prev_md5 != $data_md5) {
+        $filename = $this->fileSystem->basename($file_uri);
+        $specs_definition = $apidoc->getFieldDefinition('spec')->getItemDefinition();
+        $target_dir = $specs_definition->getSetting('file_directory');
+        $uri_scheme = $specs_definition->getSetting('uri_scheme');
+        $destination = "$uri_scheme://$target_dir/";
+
+        try {
+          $this->checkRequirements($destination);
+          $file = file_save_data($data, $destination . $filename, FILE_EXISTS_RENAME);
+
+          if (empty($file)) {
+            throw new \Exception('Could not save API Doc specification file.');
+          }
+        }
+        catch (\Exception $e) {
+          $this->log($apidoc, static::TYPE_ERROR, 'Error while saving API Doc spec file from URL on API Doc ID: %id. Error: %error', [
+            '%id' => $apidoc->id(),
+            '%error' => $e->getMessage(),
+          ], $show_messages);
+          return FALSE;
+        }
+
+        $spec_value = ['target_id' => $file->id()] + $spec_value;
+        $apidoc->set('spec', $spec_value);
+        $apidoc->set('spec_md5', $data_md5);
+        $apidoc->set('fetched_timestamp', time());
+
+        $needs_save = TRUE;
+      }
+    }
+
+    elseif ($apidoc->get('spec_file_source')->value == ApiDocInterface::SPEC_AS_FILE) {
```

Comment:
After you added a proper access check to this should not happen. I would consider throwing an exception here rather silently failing with a message.